### PR TITLE
Upload and download slices

### PR DESCRIPTION
**Pull request by g4s8**
#64 - added upload and download slices to work with `Storage` by `Key` from path

### CODE REVIEW

**Comment by paulodamaso** [comment]
@g4s8 Please assign me

**Comment by g4s8** [comment]
@0crat status

**Comment by g4s8** [comment]
@0crat elect

**Comment by 0crat** [comment]
> @0crat status ([here](https://github.com/artipie/http/issues/65#issuecomment-597491201))

@g4s8 This is what I know about this job in [CT2E6TK9B](https://www.0crat.com/p/CT2E6TK9B), as in [§32](http://www.zerocracy.com/policy.html#32):

  * The job [#65](https://github.com/artipie/http/issues/65) is [in scope](http://www.zerocracy.com/policy.html#14) for 3hr
  * The role is `REV`
  * The job is not assigned to anyone
  * The [budget](http://www.zerocracy.com/policy.html#4) is not set yet
  * These users are banned and won't be assigned:
    * @g4s8[/z](https://www.0crat.com/u/g4s8): This user reported the ticket
  * Job [footprint](https://www.0crat.com/footprint/CT2E6TK9B?q=%7Bjob%3A%27gh%3Aartipie%2Fhttp%2365%27%2Ctype%3A%7B%24not%3A%2F%5ENotify%2F%7D%7D) (restricted area)

**Comment by 0crat** [comment]
This pull request [#65](https://github.com/artipie/http/issues/65) is assigned to @fanifieiev[/z](https://www.0crat.com/u/fanifieiev), here is [why](https://www.0crat.com/footprint/CT2E6TK9B/3850a320-1e91-4e90-80a1-a4301cbf1a4f); the budget is 15 minutes, see [§4](http://www.zerocracy.com/policy.html#4); please, read [§27](http://www.zerocracy.com/policy.html#27) and when you decide to accept the changes, inform @g4s8[/z](https://www.0crat.com/u/g4s8) (the architect) right in this ticket; if you decide that this PR should not be accepted ever, also inform the architect; this [blog post](http://www.yegor256.com/2015/02/09/serious-code-reviewer.html) will help you understand what is expected from a code reviewer; there will be a monetary reward for this job

**Comment by fanifieiev** [comment]
@0crat refuse

**Comment by paulodamaso** [comment]
@g4s8 PLease assign me, I'm available

**Comment by 0crat** [comment]
> @0crat refuse ([here](https://github.com/artipie/http/issues/65#issuecomment-597654006))

@fanifieiev The user @fanifieiev[/z](https://www.0crat.com/u/fanifieiev) resigned from [#65](https://github.com/artipie/http/issues/65), please stop working. Reason for job resignation: Order was cancelled

**Comment by 0crat** [comment]
> @0crat refuse ([here](https://github.com/artipie/http/issues/65#issuecomment-597654006))

@fanifieiev Job refused in 3 hours - no penalty, see [§6](http://www.zerocracy.com/policy.html#6)

File: src/test/java/com/artipie/http/slice/SliceDownloadTest.java [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+public final class SliceDownloadTest {
+
+    @Test
+    void downloadsByKeyFromPath() throws Exception {
+        final Storage storage = new InMemoryStorage();
+        final String path = "one/two/target.txt";
+        final byte[] data = "hello".getBytes(StandardCharsets.UTF_8);
+        storage.save(new Key.From(path), new Content.From(data)).get();
+        MatcherAssert.assertThat(
+            new SliceDownload(storage).response(
```
Comment (paulodamaso) [inline src/test/java/com/artipie/http/slice/SliceDownloadTest.java]: @g4s8 I think that you need to test if `SliceDownload` is creating the response with the correct status and size too

File: src/test/java/com/artipie/http/slice/SliceUploadTest.java [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+ */
+public final class SliceUploadTest {
+
+    @Test
+    void uploadsKeyByPath() throws Exception {
+        final Storage storage = new InMemoryStorage();
+        final String hello = "Hello";
+        final byte[] data = hello.getBytes(StandardCharsets.UTF_8);
+        final String path = "uploads/file.txt";
+        MatcherAssert.assertThat(
```
Comment (paulodamaso) [inline src/test/java/com/artipie/http/slice/SliceUploadTest.java]: @g4s8 Please put each assertion in a different test method.

File: src/test/java/com/artipie/http/slice/SliceUploadTest.java [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+        );
+        MatcherAssert.assertThat(
+            new String(
+                new Remaining(
+                    Flowable.fromPublisher(storage.value(new Key.From(path)).get()).toList()
+                        .blockingGet().get(0)
+                ).bytes(),
+                StandardCharsets.UTF_8
+            ),
+            Matchers.equalTo(hello)
```
Comment (paulodamaso) [inline src/test/java/com/artipie/http/slice/SliceUploadTest.java]: @g4s8 Use `new IsEqual` here, please. Also, you can add a message to this assertion failure, even if it's the only assertion of the method: most of times is easier and clearer to read the assertion message failure than to decipher what went wrong based on test method name

File: src/main/java/com/artipie/http/slice/SliceUpload.java [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+                .thenCompose(key -> this.storage.save(key, new ContentWithSize(body, headers)))
+                .thenApply(rsp -> new RsWithStatus(RsStatus.CREATED))
+        );
+    }
+
+    /**
+     * Content with size from headers.
+     * @since 0.6
+     */
+    private static final class ContentWithSize implements Content {
```
Comment (paulodamaso) [inline src/main/java/com/artipie/http/slice/SliceUpload.java]: @g4s8 We don't need this class as static; also we could move it into its own file and detach it from `SliceUpload` so we can use it wherever we need to measure content size

**Comment by 0crat** [comment]
This pull request [#65](https://github.com/artipie/http/issues/65) is assigned to @paulodamaso[/z](https://www.0crat.com/u/paulodamaso), here is [why](https://www.0crat.com/footprint/CT2E6TK9B/124764a0-092f-4cb7-95b6-fc913c4b3605); the budget is 15 minutes, see [§4](http://www.zerocracy.com/policy.html#4); please, read [§27](http://www.zerocracy.com/policy.html#27) and when you decide to accept the changes, inform @g4s8[/z](https://www.0crat.com/u/g4s8) (the architect) right in this ticket; if you decide that this PR should not be accepted ever, also inform the architect; this [blog post](http://www.yegor256.com/2015/02/09/serious-code-reviewer.html) will help you understand what is expected from a code reviewer; there will be a monetary reward for this job

**Comment by g4s8** [comment]
@paulodamaso thanks, please check new commit

**Comment by g4s8** [comment]
@paulodamaso actually we need this PR now, so I'm merging after fixing your comments

**Comment by 0crat** [comment]
Job was finished in 19 hours, bonus for fast delivery is possible (see [§36](http://www.zerocracy.com/policy.html#36))

**Pull request merged**